content/blog/2019-12-20-Internal-Function-Memoization/index.md

```diff
@@ -0,0 +1,175 @@
+
+However, function equivalence is an undecidable problem, as that would solve the [halting problem](https://en.wikipedia.org/wiki/Halting_problem). Therefore, to approximately validate our generated programs are correct, we test the bases cases, and a few arguments for the iterative case. 
+We randomly generating a few integers between 2 and 35 as 1 is a base case, and 35 is the largest fibonacci number that still fits in a 32-bit integer.
+
+A next step to improve correctness testing would be to utilize randomized testing, but instead, determine the space of valid inputs from the input program itself.
+
+### Performance
+To measure performance, we ran derivations of the fibonacci program. We compare the execution time of the recursive program to the execution time of the iterative program.
+
+Here are the results of running the fibonacci benchmark shown above. We see that the recursive benchmark follows an $O(2^n)$ execution time trendline as expected. We also see the iterative benchmark's trendline is very linear. By the data, the iterative trendline only varies between 3 and 3.5 hundredths of milliseconds between inputs 7 and 55. This is quite small. One reason this may be is that increasing n by 1 in the iterative benchmark, likely only increasing the number of add instructions by 1.
+<img src="call.png" style="max-width: 100%" >
+
+
+When interpreting these results, we considered the difference between these two programs. 
+Consider the call tree for recursive fibonacci. As $n \longrightarrow \infty$, the number of calls in the call tree grow by $O(2^n)$.
+
+<img src="fib.png" style="max-width: 100%" >
+
+In addition to the redundant computation, the recursive fibonacci program takes up more space, as it is not tail-recursive, and therefore requires many more stack frames for each computation.
+
+Next consider Katinacci, a derivation of fibonacci with the recurrence: $$k(n) = k(n-1) + k(n-3)$$
+
+<img src="kat.png" style="max-width: 100%" >
+
+Katinacci was designed to test recurrences with offsets that were not adjacent. We noticed that Katinacci does not follow the exponential trendline quite as closely as the rest of the benchmarks. One reason this may be, is that it appraoches a base case much more quickly than the other benchmarks. Therefore, Katinacci may be a little bit lower on the graph than the exponential trendline, as shown.
+
+Finally, Henrinacci is a derivation of fibonacci with the recurrence: $$h(n) = h(n-1) + h(n-2) + h(n-3)$$
+
+<img src="rec_henri.png" style="max-width: 100%" >
+
+<img src="it_henri.png" style="max-width: 100%" >
+
+We see the recursive cases of these benchmarks follow the exponential trendline, while the iterative benchmarks look much more linear. While some of the graphs for the iterative benchmarks undulate a little, it's important to remember the time difference is quite small, and some of this could be noise from the moderately precise "time" command in bash.
```

Comment:
An easy way to answer this question would be to run the benchmark multiple times and use statistics…